### DISCUSSION TITLE
Fire event on Popover close

### DISCUSSION THREAD

**Discussion by slavachernyshev** [tailwindlabs/headlessui#1217] [Ideas]
In some cases it is useful to know, if popover was closed or not. For example, to reset logic inside PopoverPanel (tabs navigation, etc.). Like `close` event in Dialog.

```html
<Popover @close="hideHandleFn()">
```

**Comment by slavachernyshev** [comment]
For vue part it can be solved with little changes (like currently implemented in Dialog)

packages/@headlessui-vue/src/components/popover/popover.ts

```ts
export let Popover = defineComponent({
    ...
    emits: { close: (_close: boolean) => true },
    ...
    setup(props, { emit, attrs, slots }) {
    ...
    let api = {
        ...
        close(focusableElement: HTMLElement | Ref<HTMLElement | null>) {
            ...
            emit('close', false)
        }
        ...
    }
    ...
    }
})
```

**Comment by mustafaucr** [comment]
Hi for react you can use that:

```
export default function MyComponent() {
  const [popoverElement, setPopoverElement] = useState<HTMLDivElement | null>()

  useEffect(() => {
    if (!popoverElement) {
      {it means panel closed}
    }
  }, [popoverElement])

  return (
      <Popover.Panel ref={setPopoverElement} ...>
        ...
  )
}
```

**Comment by mgtcampbell** [comment]
For what it's worth if anyone else is struggling with the same thing, I was able to use a `MutationObserver` on the Popover DOM element on the `data-headlessui-state` attribute, which switches between "open" or null in order to tell when the Popover was closed by clicking outside of it.

**Comment by Sm4ster** [comment]
Since this is not supported yet, a workaround in vue is to put the content of the PopoverPanel into its own component and fire the close event from the beforeUnmount lifecycle hook.

**Comment by larxn** [comment]
You can use a ref to watch if the PopoverPanel gets closed:

**Script** tag:
```js
const popoverRef = ref(null)
watch(
  () => popoverRef.value?.el,
  (el) => {
    if (!el) {
      // someHandler() 
    }
  },
)
```

**Template** tag:
```vue
<Popover>
  <PopoverPanel ref="popoverRef" />
<Popover>
```

**Reply by Conor-Hughes** [reply]
Using the Options API I went another route. 

You can set up a ref on the <PopoverButton> which runs a function when the component updates:

```
<PopoverButton
              :ref="(el) => { runCheck(el) }"
...
```

And then in this function, check the `data-headlessui-state` attribute to see if the Popover is open:

```
methods: {
    runCheck(el)
    {
      this.$nextTick(() => {

        let state = el.$el.getAttribute("data-headlessui-state");

        if(this.previousFilterState === 'open' && !state){
          console.log('Filters closed');
        }

        this.previousFilterState = state;
      })
    },
    ...
```

Notes:
1. You need to wrap your method in this.$nextTick to avoid a recursive update loop
2. You'll need to save the previousFilter state, since this method runs on page load when the component is updated.
